### TERM INSURANCE CUSTOMER PREDICTION :

#### Import Necessary Libraries :

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

### Read File :


In [3]:
df = pd.read_csv(r"G:\Imarticus LMS\Decision Tree And Random Forest In Python\Insurance_Marketing.csv")

#### Display all rows and columns :

In [4]:
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [5]:
df.head()

Age  Job  Marital  Qualification  Default_Premium  Health_Insurance  \
0   56    3        1              0                0                 0   
1   57    7        1              1                1                 0   
2   37    7        1              1                0                 2   
3   40    0        1              0                0                 0   
4   56    7        1              1                0                 0   

   General_Insurance  Contact  Month  Last_Contact_Day  Last_Contact_Duration  \
0                  0        1      6                 1                    261   
1                  0        1      6                 1                    149   
2                  0        1      6                 1                    226   
3                  0        1      6                 1                    151   
4                  2        1      6                 1                    307   

   Contacts_During_Campaign  Previous_Contact_Days  Contacts_Before_Campaign  \
0                         1                    999                         0   
1                         1                    999                         0   
2                         1                    999                         0   
3                         1                    999                         0   
4                         1                    999                         0   

   Previous_Outcome  Employement_Rates  Price_Variation  \
0                 1                1.1        94.159686   
1                 1                1.1        93.374479   
2                 1                1.1        94.100629   
3                 1                1.1        92.772613   
4                 1                1.1        93.241248   

   Consumer_Confidence_Index  Insurance_Rate  No_Employees  Client_Subscribed  
0                 -38.173679        3.543298   5074.493963                  0  
1                 -33.312428        6.013585   5125.221264                  0  
2                 -51.947349        2.784063   5205.455324                  0  
3                 -35.624584        4.459237   5096.139930                  0  
4                 -33.492147        3.988744   5112.681264                  0

####  Get the number of Rows and Columns :


In [6]:
df.shape

(41188, 21)

#### Check for any misssing null values in the data :


In [7]:
df.isna().sum()

Age                          0
Job                          0
Marital                      0
Qualification                0
Default_Premium              0
Health_Insurance             0
General_Insurance            0
Contact                      0
Month                        0
Last_Contact_Day             0
Last_Contact_Duration        0
Contacts_During_Campaign     0
Previous_Contact_Days        0
Contacts_Before_Campaign     0
Previous_Outcome             0
Employement_Rates            0
Price_Variation              0
Consumer_Confidence_Index    0
Insurance_Rate               0
No_Employees                 0
Client_Subscribed            0
dtype: int64

#### Display Column Names :

In [8]:
df.columns

Index(['Age', 'Job', 'Marital', 'Qualification', 'Default_Premium',
       'Health_Insurance', 'General_Insurance', 'Contact', 'Month',
       'Last_Contact_Day', 'Last_Contact_Duration', 'Contacts_During_Campaign',
       'Previous_Contact_Days', 'Contacts_Before_Campaign', 'Previous_Outcome',
       'Employement_Rates', 'Price_Variation', 'Consumer_Confidence_Index',
       'Insurance_Rate', 'No_Employees', 'Client_Subscribed'],
      dtype='object')

#### Get Datatypes :

In [9]:
df.dtypes

Age                            int64
Job                            int64
Marital                        int64
Qualification                  int64
Default_Premium                int64
Health_Insurance               int64
General_Insurance              int64
Contact                        int64
Month                          int64
Last_Contact_Day               int64
Last_Contact_Duration          int64
Contacts_During_Campaign       int64
Previous_Contact_Days          int64
Contacts_Before_Campaign       int64
Previous_Outcome               int64
Employement_Rates            float64
Price_Variation              float64
Consumer_Confidence_Index    float64
Insurance_Rate               float64
No_Employees                 float64
Client_Subscribed              int64
dtype: object

#### Split the Data into train and test :

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
df_train , df_test = train_test_split(df , test_size = .2 ,random_state = 555)

In [12]:
df_train_x = df_train.iloc[: , 0:-1]
df_train_y = df_train.iloc[: , -1]

In [13]:
df_test_x = df_test.iloc[: , 0:-1]
df_test_y = df_test.iloc[: , -1]

In [14]:
print(df_train_x.shape)
print(df_train_y.shape)

(32950, 20)
(32950,)


In [15]:
print(df_test_x.shape)
print(df_test_y.shape)

(8238, 20)
(8238,)


#### Now Apply Decision Tree Classifier  :

#### Model Formation :

In [17]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dt.fit(df_train_x , df_train_y)

DecisionTreeClassifier()

#### Prediction :

In [19]:
pred_dt  = dt.predict(df_test_x )
pred_dt

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

#### Confusion Matrix :

In [20]:
from sklearn.metrics import confusion_matrix
tab_dt =  confusion_matrix(pred_dt   , df_test_y )
tab_dt

array([[6836,  488],
       [ 452,  462]], dtype=int64)

#### Efficiency :

In [21]:
tab_dt.diagonal().sum() * 100  / tab_dt.sum()


88.58946346200534

# .................................................................................................

#### Now Apply Random Forest :


#### Model Formation :

In [22]:
from sklearn.ensemble import RandomForestClassifier
rfc  = RandomForestClassifier()


In [23]:
rfc  = RandomForestClassifier(n_estimators  = 10)
rfc.fit(df_train_x , df_train_y)

RandomForestClassifier(n_estimators=10)

#### Prediction :

In [24]:
pred_rfc  = rfc.predict(df_test_x )
pred_rfc

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

#### Confusion Matrix :

In [25]:
tab1_dt  = confusion_matrix(pred_rfc , df_test_y)
tab1_dt

array([[7112,  604],
       [ 176,  346]], dtype=int64)

#### Efficiency :

In [26]:
tab1_dt.diagonal().sum() * 100  / tab1_dt.sum()

90.53168244719592

# ........................................................................................................

#### Now Apply ADA Boost :

#### Model Formation :


In [27]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier


In [28]:
dt = DecisionTreeClassifier()

In [29]:
abc = AdaBoostClassifier( dt   ,  10)


In [30]:
abc.fit(df_train_x , df_train_y)


AdaBoostClassifier(base_estimator=DecisionTreeClassifier(), n_estimators=10)

#### Prediction :


In [31]:
pred_abc = abc.predict(df_test_x)
pred_abc

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

#### Confusion Matrix :


In [32]:
tab_abc = confusion_matrix(pred_abc , df_test_y)
tab_abc

array([[6837,  481],
       [ 451,  469]], dtype=int64)

#### Efficiency :

In [33]:
tab_abc.diagonal().sum() * 100  / tab_abc.sum()

88.68657441126487

# ......................................................................................

### Conclusion :

#### Here I have built various types of model such as Decision Tree Classifier , Random Forest Classifier and ADA Boost Model .

### Accuracy / Efficiency :

#### Decision Tree Classifier         =  88.58946346200534

#### Random Forest Classifier      =  90.53168244719592

#### ADA Boost                               =  88.68657441126487


#### By cmparing above Accuracies / Efficiencies , I have selected Random Forest Classifier for Model Formation , because it has highest Accuracy , Which is equal to  88.58946346200534 .


# .................................... END...................................................